src/rtp_receiver.rs

```diff
@@ -39,228 +26,120 @@ pub struct Options {
-                ffmpeg_next::codec::Id::H264 => compositor_pipeline::pipeline::decoder::Codec::H264,
-                _ => unimplemented!(),
-            },
+            codec: compositor_pipeline::pipeline::decoder::Codec::H264,
         }
     }
 }
 
-#[derive(Debug, thiserror::Error)]
-pub enum InitError {
-    #[error(transparent)]
-    FfmpegError(#[from] ffmpeg_next::Error),
-    #[error(transparent)]
-    IoError(#[from] std::io::Error),
-}
-
 impl RtpReceiver {
-    fn run(
-        port: u16,
-        should_close: Receiver<()>,
-        packet_sender: Sender<Packet>,
-        decoder_params_sender: Sender<
-            Result<DecoderParameters, Box<dyn std::error::Error + Send + Sync + 'static>>,
-        >,
+    async fn rtp_receiver(
+        socket: net::UdpSocket,
+        packets_tx: Sender<rtp::packet::Packet>,
+        should_close_rx: Receiver<()>,
     ) {
-        let sdp_filepath = PathBuf::from(format!("/tmp/sdp_input_{}.sdp", port));
-        let sdp_file_result = File::create(&sdp_filepath)
-            .map_err(InitError::IoError)
-            .and_then(|mut file| {
-                file.write_all(
-                    format!(
-                        "\
-                        v=0\n\
-                        o=- 0 0 IN IP4 127.0.0.1\n\
-                        s=No Name\n\
-                        c=IN IP4 127.0.0.1\n\
-                        m=video {} RTP/AVP 96\n\
-                        a=rtpmap:96 H264/90000\n\
-                        a=fmtp:96 packetization-mode=1\n\
-                        a=rtcp-mux\n\
-                    ",
-                        port
-                    )
-                    .as_bytes(),
-                )
-                .map_err(InitError::IoError)
-            });
+        let mut buffer = BytesMut::zeroed(65536);
 
-        if let Err(e) = sdp_file_result {
-            decoder_params_sender.send(Err(e.into())).unwrap();
-            return;
-        }
-
-        // careful: moving the input context in any way will cause ffmpeg to segfault
-        // I do not know why this happens
-        let input_ctx = match input_with_dictionary_and_interrupt(
-            &sdp_filepath,
-            Dictionary::from_iter([("protocol_whitelist", "file,udp,rtp")]),
-            || should_close.try_recv().is_ok(),
-        ) {
-            Ok(i) => i,
-            Err(e) => {
-                decoder_params_sender
-                    .send(Err(Box::new(InitError::FfmpegError(e))))
-                    .unwrap();
-                return;
-            }
-        };
+        loop {
+            let received_bytes = future::or(
+                async {
+                    should_close_rx.recv().await.unwrap();
+                    None
+                },
+                async {
+                    let (len, _) = socket.recv_from(&mut buffer).await.unwrap();
+                    Some(len)
+                },
+            )
+            .await;
+
+            match received_bytes {
```

Comment:
```suggestion
            let Some(received_bytes) = received_bytes else {
                return
            }
```

clearer intention + less indentation in the rest of the code.

technically you could do break, but I think return shows intention better

Reply:
I want to address this in the following PRs, so I'm leaving this as-is for now.